### PR TITLE
[RFC]Add yanks provider

### PR DESCRIPTION
**Pull request by ratheesh**
Add yanks provider
This is largely based on https://github.com/sgur/ctrlp-extensions.vim

This should address #76 

I am new to vim scripting. Suggestions are welcome!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,6 +94,7 @@ Command                                | List                               | Re
 `Clap lines`                           | Lines in the loaded buffers        | _none_
 `Clap marks`                           | Marks                              | _none_
 `Clap tags`                            | Tags in the current buffer         | **[vista.vim][vista.vim]**
+`Clap yanks`                           | Yank stack of the current vim session    | ViM(8.0.1394+)/NeoVim(0.3.0+)
```
Comment (liuchengxu) [inline README.md]: I believe the requirement has been covered by clap itself.

Comment (ratheesh) [reply README.md]: Yes, adjusted accordingly.

File: autoload/clap/provider/yanks.vim [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+endfunction
+
+function! s:yanks_sink(selected) abort
+    call setreg('"', a:selected)
+    normal! ""p
+endfunction
+
+let s:yanks.sink = function('s:yanks_sink')
+
+function! s:yanks.on_enter() abort
```
Comment (liuchengxu) [inline autoload/clap/provider/yanks.vim]: The `on_enter` hook is unccessary as the collection is done via the TextYankPost autocmd already. It's often used in this plugin to initialize the filetype of display buffer to get the syntax highlight.

File: autoload/clap/provider/yanks.vim [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+endfunction
+
+function! s:yanks_collect() abort
+    let yankstr = getreg('"')
+
+    if !empty(s:yanklist) && yankstr == s:yanklist[0]
+        return
+    endif
+
+    call s:remove_duplicated(s:yanklist, yankstr)
```
Comment (liuchengxu) [inline autoload/clap/provider/yanks.vim]: Could be simpified using `filter`, `:h filter()`.

File: autoload/clap/provider/yanks.vim [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+function! s:yanks_sink(selected) abort
+    call setreg('"', a:selected)
+    normal! ""p
+endfunction
+
+let s:yanks.sink = function('s:yanks_sink')
+
+function! s:yanks.on_enter() abort
+    if !exists('##TextYankPost')
```
Comment (liuchengxu) [inline autoload/clap/provider/yanks.vim]: This check is also needless as the requirement of clap ensures the feature exists.

File: autoload/clap/provider/yanks.vim (hunk 2 above)
Comment (ratheesh) [reply autoload/clap/provider/yanks.vim]: `on_enter` hook is removed

File: autoload/clap/provider/yanks.vim (hunk 3 above)
Comment (ratheesh) [reply autoload/clap/provider/yanks.vim]: replaced function with `filter()` call as suggested.

File: autoload/clap/provider/yanks.vim (hunk 4 above)
Comment (ratheesh) [reply autoload/clap/provider/yanks.vim]: Removed the redundant part.

File: plugin/yanks.vim [hunk 5]
```diff
@@ -0,0 +1,16 @@
+" vim-clap - Modern interactive filter and dispatcher
```
Comment (liuchengxu) [inline plugin/yanks.vim]: Is it better to put this files in plug/clap.vim?

```vim
if get(g:, 'clap_enable_yanks_provider', 1)
  augroup ClapYanks
    autocmd!
    autocmd VimEnter * call clap#provider#yanks#init()
  augroup END
endif
```

Comment (ratheesh) [reply plugin/yanks.vim]: Done!

File: autoload/clap/provider/yanks.vim [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+" Based on : https://github.com/sgur/ctrlp-extensions.vim
+
+let s:save_cpo = &cpo
+set cpo&vim
+
+let s:yanklist = []
+let s:yanks = {}
+
+function! s:yanks_collect() abort
+    let yankstr = getreg('"')
```
Comment (liuchengxu) [inline autoload/clap/provider/yanks.vim]: Use 2 spaces for every level of indentation in Vim code.

File: autoload/clap/provider/yanks.vim [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    if !empty(s:yanklist) && yankstr == s:yanklist[0]
+        return
+    endif
+
+    call filter(s:yanklist, 'v:val != yankstr')
+    call insert(s:yanklist, yankstr)
+endfunction
+
+function! clap#provider#yanks#yanks_init() abort
+    augroup clap_yanklist
```
Comment (liuchengxu) [inline autoload/clap/provider/yanks.vim]: The augroup name should be CamelCase, e.g, ClapYanksCollect.

File: autoload/clap/provider/yanks.vim [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    if !empty(s:yanklist) && yankstr == s:yanklist[0]
+        return
+    endif
+
+    call filter(s:yanklist, 'v:val != yankstr')
+    call insert(s:yanklist, yankstr)
+endfunction
+
+function! clap#provider#yanks#yanks_init() abort
```
Comment (liuchengxu) [inline autoload/clap/provider/yanks.vim]: Personally I would rename to `clap#provider#yanks#init()` :(.

**Comment by liuchengxu** [comment]
Overall good, just some nits. And I also have some ideas for the future improments:

- record the time of inserted history.
- add limit to the yanks history, e.g., keep up to 100 entries.
- save the yanks history to a local file so that we are able to use them when reopening vim.

File: autoload/clap/provider/yanks.vim [hunk 9]
```diff
@@ -0,0 +1,46 @@
+" Author: Ratheesh S <ratheeshreddy@gmail.com>
+" Description: List the recently yanked/deleted lines
+" Based on : https://github.com/sgur/ctrlp-extensions.vim
+
+let s:save_cpo = &cpo
+set cpo&vim
+
+let s:yanklist = []
```
Comment (liuchengxu) [inline autoload/clap/provider/yanks.vim]: `s:yanks_history` is a better name?

File: autoload/clap/provider/yanks.vim (hunk 6 above)
Comment (liuchengxu) [inline autoload/clap/provider/yanks.vim]: `last_yanked` a better name?

File: plugin/clap.vim [hunk 10]
```diff
@@ -19,3 +19,11 @@ augroup ClapBuffers
   autocmd BufWinEnter,WinEnter * let g:__clap_buffers[bufnr('')] = reltimefloat(reltime())
   autocmd BufDelete * silent! call remove(g:__clap_buffers, expand('<abuf>'))
 augroup END
+
+" yanks provider
+if get(g:, 'clap_enable_yanks_provider', 1)
+  augroup ClapYanks
+    autocmd!
+    autocmd VimEnter * call clap#provider#yanks#init()
+  augroup END
```
Comment (liuchengxu) [inline plugin/clap.vim]: We could avoid sourcing the yanks provider on VimEnter by introducing a global variable, although the sourcing is actually cheap :(.

```vim
if get(g:, 'clap_enable_yanks_provider', 1)
  let g:__clap_last_yanked = getreg('"')
  augroup ClapYanks
    autocmd!
    autocmd TextYankPost * call clap#provider#yanks#collect()
  augroup END
endif
```

Initialize `s:yank_history` with `g:__clap_last_yanked`:


```vim
let s:yank_history = exists('g:__clap_last_yanked') ? [g:__clap_last_yanked] : []
```

Expose `s:yanks_collect()` to `clap#provider#yanks#collect()`, and then we can remove `clap#provider#yanks#init()`.

Comment (ratheesh) [reply plugin/clap.vim]: `getreg()` in the above code executes before reading of `viminfo/shada` file before the VimEnter event.  As a result, yank provider starts with an empty list(even if register `"` is nonempty). Hence we may need to live with VimEnter autocmd for now. 

I have exposed `s:yanks_collect()` to `clap#provider#yanks#collect()`
I have implemented a max. yank entries through global variable(Default:20)

**Comment by ratheesh** [comment]
> record the time of inserted history.

What is the use case scenario of the yank timestamp?

> add a limit to the yanks history, e.g., keep up to 100 entries.

This is included in current PR(latest commit)

> save the yanks history to a local file so that we are able to use them when reopening vim.

We can do this in a separate PR.

**Comment by alerque** [comment]
Being a clipboard manager is a complicated problem, and not one you will find a lot of agreement on. I **do not want** my yanks being written to a file. I'm pretty against them hitting my disk even in temp space actually. Please don't hold up this PR for stuff like that. Having easy visual access to yanks will be awesome. Trying to much about with how yanks work in the first place should be out of scope for this.

File: autoload/clap/provider/yanks.vim [hunk 11]
```diff
@@ -0,0 +1,53 @@
+" Author: Ratheesh S <ratheeshreddy@gmail.com>
+" Description: List the recently yanked/deleted lines
+" Based on : https://github.com/sgur/ctrlp-extensions.vim
+
+let s:save_cpo = &cpo
+set cpo&vim
+
+let s:yank_history = []
+let s:yanks = {}
+let s:max_yanks = get(g:, 'clap_yanks_max_entries', 20)
```
Comment (liuchengxu) [inline autoload/clap/provider/yanks.vim]: Please use `clap_provider_yanks_max_entries`.

**Comment by liuchengxu** [comment]
Well, the ideas above are merely my quick thoughts, they don't have to be done in this PR or even later actually :(.

>What is the use case scenario of the yank timestamp?

I just thought it would be interesting to know about the time of that yank operation. It's essentially unimportant.

>getreg() in the above code executes before reading of viminfo/shada file before the VimEnter event.
Good to know! Then I'm ok with that.

File: autoload/clap/provider/yanks.vim (hunk 11 above)
Comment (liuchengxu) [reply autoload/clap/provider/yanks.vim]: Also add the doc about this option in doc/clap.txt.

File: plugin/clap.vim [hunk 12]
```diff
@@ -19,3 +19,11 @@ augroup ClapBuffers
   autocmd BufWinEnter,WinEnter * let g:__clap_buffers[bufnr('')] = reltimefloat(reltime())
   autocmd BufDelete * silent! call remove(g:__clap_buffers, expand('<abuf>'))
 augroup END
+
+" yanks provider
+if get(g:, 'clap_enable_yanks_provider', 1)
```
Comment (liuchengxu) [inline plugin/clap.vim]: And add the doc about this option in doc/clap.txt.

**Comment by liuchengxu** [comment]
Merged, thanks! Nice job @ratheesh  👍

**Pull request merged**